File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    int[][] pacific;
    int[][] atlantic;
    int[] dr = new int[]{1, -1, 0, 0};
    int[] dc = new int[]{0, 0, 1, -1};

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        pacific = new int[heights.length][heights[0].length];
        atlantic = new int[heights.length][heights[0].length];

        for (int i = 0; i < heights.length; i++) {
            dfs(heights, pacific, i, 0);
        }

        for (int i = 0; i < heights.length; i++) {
            dfs(heights, atlantic, i, heights[0].length - 1);
        }

        for (int j = 0; j < heights[0].length; j++) {
            dfs(heights, pacific, 0, j);
        }

        for (int j = 0; j < heights[0].length; j++) {
            dfs(heights, atlantic, heights.length - 1, j);
        }

        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < heights.length; i++) {
            for (int j = 0; j < heights[0].length; j++) {
                if (pacific[i][j] == 1 && atlantic[i][j] == 1) {
                    Integer[] point = new Integer[]{i, j};
                    res.add(Arrays.asList(point));
                }
            }
        }

        return res;
       
    }

    private void dfs(int[][] heights, int[][] used, int row, int col) {
        if (used[row][col] == 1) {
            return;
        }

        used[row][col] = 1;

        for (int k = 0; k < 4; k++) {
            int newRow = row + dr[k];
            int newCol = col + dc[k];

            if (newRow >= 0 && newRow < heights.length && newCol >= 0 && newCol < heights[0].length && heights[newRow][newCol] >= heights[row][col]) {
                dfs(heights, used, newRow, newCol);
            }
        }
    }
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:

    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        dr, dc = [0, 0, 1, -1], [1, -1, 0, 0]

        m, n = len(heights), len(heights[0])

        pacific = [[False] * n for _ in range(m)]
        atlantic = [[False] * n for _ in range(m)]
        
        def dfs(row, col, visited):
            visited[row][col] = True

            for k in range(4):
                x, y = row + dr[k], col + dc[k]

                if 0 <= x < m and 0 <= y < n and not visited[x][y]:
                    if heights[x][y] >= heights[row][col]:
                        visited[x][y] = True
                        dfs(x, y, visited)

        for i in range(m):
            dfs(i, 0, pacific)
            dfs(i, n - 1, atlantic)
        
        for j in range(n):
            dfs(0, j, pacific)
            dfs(m - 1, j, atlantic)

        res = [[i, j] for i in range(m) for j in range(n) if pacific[i][j] and atlantic[i][j]]
        
        return res

    

        